### PR TITLE
Fix S1751: "Unconditional jump statements" should not raise if a conditional continue is detected

### LINKED ISSUE #345: Fix S1751: "Unconditional jump statements" should not raise if a conditional continue is detected
[RSPEC-1751](https://jira.sonarsource.com/browse/RSPEC-1751)

Do not raise if the loop also contains a conditional "continue" statement.

### PR DESCRIPTION
**Pull request by michalb-sonar**
Fix #345

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs [hunk 1]
```diff
@@ -93,13 +70,97 @@ protected override void Initialize(SonarAnalysisContext context)
             (statement as ReturnStatementSyntax)?.ReturnKeyword.ToString() ??
             (statement as ThrowStatementSyntax)?.ThrowKeyword.ToString();
 
-        private static bool IsWithinMethodPropertyOrLambda(SyntaxNode node) =>
-            !node.IsAnyKind(methodOrPropertyDeclarations);
+        private class FooWalker : CSharpSyntaxWalker
+        {
+            private readonly SyntaxNodeAnalysisContext context;
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs]: MAJOR Remove the 'context' field and declare it as a local variable in the relevant methods. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1450)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs [hunk 2]
```diff
@@ -93,13 +70,95 @@ protected override void Initialize(SonarAnalysisContext context)
...
+
+                ruleViolations.AddRange(unconditionalContinues);
+
+                if (unconditionalTerminates.Any() &&
+                    !conditionalContinues.Any())
+                {
+                    ruleViolations.AddRange(unconditionalTerminates);
+                }
+
+                return ruleViolations;
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs]: Isn't the content of this method basically `unconditionalTerminates.Except(conditionalContinues).ToList()`? I am not really sure what this method does...

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs [hunk 3]
```diff
@@ -93,13 +70,95 @@ protected override void Initialize(SonarAnalysisContext context)
             (statement as ReturnStatementSyntax)?.ReturnKeyword.ToString() ??
             (statement as ThrowStatementSyntax)?.ThrowKeyword.ToString();
 
-        private static bool IsWithinMethodPropertyOrLambda(SyntaxNode node) =>
-            !node.IsAnyKind(methodOrPropertyDeclarations);
+        private class FooWalker : CSharpSyntaxWalker
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs]: The name of this class is not good.

Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnconditionalJumpStatement.cs]: Does this walker support nested loops? Something like:

    while (true)
    {
        while (true)
        {
            if (something) continue;
        }
        break;
    }

**Pull request merged**